tests/test_main.py

```diff
@@ -0,0 +1,318 @@
+    requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                      json=get_matching_helper)
+
+    # self-enrolment-methods <SOURCE_ID> <TARGET_ID>
+    result = runner.invoke(cli, ['self-enrolment-methods',
+                                 '2', '3'],
+                           env={'MOODLE_URL': TEST_MOODLE_URL,
+                                'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+    assert result.exit_code == 0
+
+
+def test_set_self_enrolment_method_key(requests_mock):
+    runner = CliRunner()
+    requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                       json=post_matching_helper)
+
+    # set-self-enrolment-methods-key <ENROL_ID> <ENROL_KEY>
+    result = runner.invoke(cli, ['set-self-enrolment-method-key',
+                                 '2', 'abc123'],
+                           env={'MOODLE_URL': TEST_MOODLE_URL,
+                                'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+    assert result.exit_code == 0
+
+
+def test_course_bulk_setup_error(requests_mock, tmp_path):
+    """This exercises an error that can be caused by invalid email inputs
+    in the course-bulk-setup CLI command"""
+    runner = CliRunner()
+    with runner.isolated_filesystem(temp_dir=tmp_path):
+        requests_mock.get(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                          json=get_matching_helper)
+        requests_mock.post(f'{TEST_MOODLE_URL}{MOODLE_WEBSERVICE_PATH}',
+                           json=post_matching_helper)
+
+        f = open('test.csv', 'w')
+        writer = csv.DictWriter(
+                f,
+                utils.course_bulk_input_csv_fieldnames()
+            )
+        writer.writeheader()
+        writer.writerows([{utils.CSV_INST_FNAME: 'Tom',
+                           utils.CSV_INST_LNAME: 'Michaels',
+                           utils.CSV_INST_EMAIL: 'tommichaels@gmail.com',
+                           utils.CSV_INST_AUTH: 'manual',
+                           utils.CSV_COURSE_NAME: 'Algebra',
+                           utils.CSV_COURSE_SHORTNAME: 'a1',
+                           utils.CSV_COURSE_CATEGORY: 'MISC'},
+                          {utils.CSV_INST_FNAME: 'Aliza',
+                           utils.CSV_INST_LNAME: 'Brown',
+                           utils.CSV_INST_EMAIL: 'invalid@gmail.com',
+                           utils.CSV_INST_AUTH: 'manual',
+                           utils.CSV_COURSE_NAME: 'English',
+                           utils.CSV_COURSE_SHORTNAME: 'e1',
+                           utils.CSV_COURSE_CATEGORY: 'MISC'}])
+        f.close()
+
+        result = runner.invoke(cli, ['course-bulk-setup',
+                                     '2', 'test.csv', 'output.csv'],
+                               env={'MOODLE_URL': TEST_MOODLE_URL,
+                                    'MOODLE_TOKEN': TEST_MOODLE_TOKEN})
+
+        assert result.exit_code == 0
+        assert os.stat("output.csv").st_size != 0
```

Comment:
Similar to the other `course-bulk-setup` test case, it'd probably be good to parse the `output.csv` here using `csv.DictReader` and confirming it has a single row of data.